core/src/impl/KokkosExp_IterateTileGPU.hpp

```diff
@@ -1151,2570 +881,68 @@ struct DeviceIterateTile<5, PolicyType, Functor, Tag> {
-      const EmulateCUDADim3<index_type> blockIdx_,
-      const EmulateCUDADim3<index_type> threadIdx_)
-      : m_policy(policy_),
-        m_func(f_),
-        m_v(v_),
-        gridDim(gridDim_),
-        blockIdx(blockIdx_),
-        threadIdx(threadIdx_) {}
-#else
-  KOKKOS_IMPL_DEVICE_FUNCTION DeviceIterateTile(const PolicyType& policy_,
-                                                const Functor& f_,
-                                                value_type* v_)
-      : m_policy(policy_), m_func(f_), m_v(v_) {}
-#endif
-
-  static constexpr index_type max_blocks = 65535;
-
-  KOKKOS_IMPL_DEVICE_FUNCTION
-  void exec_range() const {
-    if (static_cast<index_type>(blockIdx.x) < m_policy.m_num_tiles &&
-        static_cast<index_type>(threadIdx.y) < m_policy.m_prod_tile_dims) {
-      index_type m_offset[PolicyType::rank];  // tile starting global id offset
-      index_type
-          m_local_offset[PolicyType::rank];  // tile starting global id offset
-
-      for (index_type tileidx = static_cast<index_type>(blockIdx.x);
-           tileidx < m_policy.m_num_tiles; tileidx += gridDim.x) {
-        index_type tile_idx =
-            tileidx;  // temp because tile_idx will be modified while
-                      // determining tile starting point offsets
-        index_type thrd_idx = static_cast<index_type>(threadIdx.y);
-        bool in_bounds      = true;
-
-        // LL
-        if (PolicyType::inner_direction == PolicyType::Left) {
-          for (int i = 0; i < PolicyType::rank; ++i) {
-            m_offset[i] =
-                (tile_idx % m_policy.m_tile_end[i]) * m_policy.m_tile[i] +
-                m_policy.m_lower[i];
-            tile_idx /= m_policy.m_tile_end[i];
+template <typename T>
+struct is_array_type<T[]> : std::true_type {
+  using value_type = T;
+};
 
-            // tile-local indices identified with threadIdx.y
-            m_local_offset[i] = (thrd_idx % m_policy.m_tile[i]);
-            thrd_idx /= m_policy.m_tile[i];
+// ------------------------------------------------------------------ //
 
-            m_offset[i] += m_local_offset[i];
-            if (!(m_offset[i] < m_policy.m_upper[i] &&
-                  m_local_offset[i] < m_policy.m_tile[i])) {
-              in_bounds &= false;
-            }
-          }
-          if (in_bounds) {
-            m_func(m_offset[0], m_offset[1], m_offset[2], m_offset[3],
-                   m_offset[4], m_offset[5], m_v);
-          }
-        }
-        // LR
-        else {
-          for (int i = PolicyType::rank - 1; i >= 0; --i) {
-            m_offset[i] =
-                (tile_idx % m_policy.m_tile_end[i]) * m_policy.m_tile[i] +
-                m_policy.m_lower[i];
-            tile_idx /= m_policy.m_tile_end[i];
+template <typename T>
+using value_type_storage_t =
+    typename std::conditional<is_array_type<T>::value, std::decay<T>,
+                              std::add_lvalue_reference<T> >::type::type;
```

Comment:
```suggestion
    typename std::conditional_t<is_array_type<T>::value, std::decay<T>,
                                std::add_lvalue_reference<T> >::type;
```

Reply:
Done in [`5e59404` (#3624)](https://github.com/kokkos/kokkos/pull/3624/commits/5e5940405575404039658fa8a5c55724d7a2dccd).